File: GPU/PipelineFactory.cs
// ────── ╔╗                                                                                    GPU
// ╔═╦╦═╦╦╬╣ PipelineFactory.cs
// ║║║║╬║╔╣║ Pre-compiles all WebGPU render pipelines at initialization time
// ╚╩═╩═╩╝╚╝ ───────────────────────────────────────────────────────────────────────────────────────
namespace Nori;
using System.IO;
using System.Reflection;
using System.Collections.Generic;

#region enum EPipeline ----------------------------------------------------------------------------------
/// <summary>Identifies a pre-compiled WebGPU render pipeline</summary>
public enum EPipeline {
   /// <summary>2D anti-aliased line via instanced quads</summary>
   Line2D,
   /// <summary>3D anti-aliased line via instanced quads</summary>
   Line3D,
   /// <summary>2D Bezier curve (CPU pre-tessellated) via instanced quads</summary>
   Bezier2D,
   /// <summary>2D dashed line with line-type texture</summary>
   DashLine2D,
   /// <summary>2D point via instanced quads</summary>
   Point2D,
   /// <summary>3D point via instanced quads</summary>
   Point3D,
   /// <summary>2D filled triangles (flat color)</summary>
   Triangle2D,
   /// <summary>2D filled quads (CPU-expanded to triangles, flat color)</summary>
   Quad2D,
   /// <summary>3D wireframe (same shader as Line3D)</summary>
   BlackLine,
   /// <summary>3D stippled line</summary>
   GlassLine,
   /// <summary>3D Gouraud shading (per-vertex lighting)</summary>
   Gourad,
   /// <summary>3D Phong shading (per-fragment lighting)</summary>
   Phong,
   /// <summary>3D Phong with pink back-face highlighting</summary>
   PhongPink,
   /// <summary>3D pick buffer (entity ID as color)</summary>
   Pick,
   /// <summary>3D stippled surface (glass effect)</summary>
   Glass,
   /// <summary>3D flat-facet shading (provoking vertex normal)</summary>
   FlatFacet,
   /// <summary>Pixel-space text rendering</summary>
   TextPx,
   /// <summary>2D world-space text rendering</summary>
   Text2D,
   /// <summary>3D world-space text rendering</summary>
   Text3D,
   /// <summary>Stencil write pass for triangle-fan fill (invert stencil bit)</summary>
   TriFanStencil,
   /// <summary>Stencil test pass for triangle-fan fill (draw where stencil set)</summary>
   TriFanCover,
}
#endregion

#region class PipelineFactory ----------------------------------------------------------------------------
/// <summary>Pre-compiles all WebGPU render pipelines at initialization time</summary>
/// In WebGPU, pipelines are immutable and pre-compiled (unlike OpenGL where state
/// is set dynamically). This factory loads all WGSL shaders from embedded resources,
/// creates bind group layouts, and builds every pipeline the renderer needs.
public unsafe class PipelineFactory : IDisposable {
   // Properties ---------------------------------------------------------------
   /// <summary>The uniform-only bind group layout (Group 0: uniform buffer at binding 0)</summary>
   public BindGroupLayout* UniformLayout => mUniformLayout;

   /// <summary>The textured bind group layout (Group 0: uniform + texture + sampler)</summary>
   public BindGroupLayout* TexturedLayout => mTexturedLayout;

   // Methods ------------------------------------------------------------------
   /// <summary>Create and compile all render pipelines for the given device</summary>
   public static PipelineFactory Create (GPUDevice gpu) {
      PipelineFactory factory = new () { mGPU = gpu };
      factory.Init ();
      return factory;
   }

   /// <summary>Retrieve a pre-compiled render pipeline by type</summary>
   public GPUPipeline Get (EPipeline pipeline) => mPipelines[pipeline];

   /// <summary>Release all pipelines, shader modules, and bind group layouts</summary>
   public void Dispose () {
      foreach (GPUPipeline pipe in mPipelines.Values) pipe.Dispose ();
      mPipelines.Clear ();
      foreach (GPUShaderModule sm in mShaders.Values) sm.Dispose ();
      mShaders.Clear ();
      if (mUniformLayout != null) { mGPU.Api.BindGroupLayoutRelease (mUniformLayout); mUniformLayout = null; }
      if (mTexturedLayout != null) { mGPU.Api.BindGroupLayoutRelease (mTexturedLayout); mTexturedLayout = null; }
      GC.SuppressFinalize (this);
   }

   // Implementation -----------------------------------------------------------
   void Init () {
      LoadShaders ();
      CreateBindGroupLayouts ();
      CreateAllPipelines ();
   }

   // Loads all WGSL shader sources from embedded resources and compiles them
   // into GPUShaderModule objects. Each .wgsl file becomes a module keyed by
   // its filename without extension (e.g., "Line2D", "Flat2D").
   void LoadShaders () {
      Assembly asm = typeof (PipelineFactory).Assembly;
      string[] names = asm.GetManifestResourceNames ();
      foreach (string name in names) {
         if (!name.EndsWith (".wgsl")) continue;
         // Resource name format: Nori.GPU.Shaders.Line2D.wgsl (or similar)
         string key = Path.GetFileNameWithoutExtension (name);
         using Stream stream = asm.GetManifestResourceStream (name)!;
         using StreamReader reader = new (stream);
         string source = reader.ReadToEnd ();
         mShaders[key] = GPUShaderModule.Create (mGPU, source, label: key);
      }
   }

   // Creates the two bind group layouts used across all pipelines:
   //   - Uniform layout: binding(0) = uniform buffer
   //   - Textured layout: binding(0) = uniform buffer, binding(1) = texture, binding(2) = sampler
   void CreateBindGroupLayouts () {
      // Uniform-only layout (used by most pipelines)
      BindGroupLayoutEntry uniformEntry = new () {
         Binding = 0,
         Visibility = ShaderStage.Vertex | ShaderStage.Fragment,
         Buffer = new BufferBindingLayout {
            Type = BufferBindingType.Uniform,
            HasDynamicOffset = false,
            MinBindingSize = 0
         }
      };
      BindGroupLayoutDescriptor uniformDesc = new () {
         EntryCount = 1,
         Entries = &uniformEntry
      };
      mUniformLayout = mGPU.Api.DeviceCreateBindGroupLayout (mGPU.Device, &uniformDesc);

      // Textured layout (used by DashLine2D, TextPx, Text2D, Text3D)
      BindGroupLayoutEntry* texEntries = stackalloc BindGroupLayoutEntry[3];
      texEntries[0] = new BindGroupLayoutEntry {
         Binding = 0,
         Visibility = ShaderStage.Vertex | ShaderStage.Fragment,
         Buffer = new BufferBindingLayout {
            Type = BufferBindingType.Uniform,
            HasDynamicOffset = false,
            MinBindingSize = 0
         }
      };
      texEntries[1] = new BindGroupLayoutEntry {
         Binding = 1,
         Visibility = ShaderStage.Fragment,
         Texture = new TextureBindingLayout {
            SampleType = TextureSampleType.Float,
            ViewDimension = TextureViewDimension.Dimension2D,
            Multisampled = false
         }
      };
      texEntries[2] = new BindGroupLayoutEntry {
         Binding = 2,
         Visibility = ShaderStage.Fragment,
         Sampler = new SamplerBindingLayout {
            Type = SamplerBindingType.Filtering
         }
      };
      BindGroupLayoutDescriptor texDesc = new () {
         EntryCount = 3,
         Entries = texEntries
      };
      mTexturedLayout = mGPU.Api.DeviceCreateBindGroupLayout (mGPU.Device, &texDesc);
   }

   // Creates all 21 render pipelines from Index.txt using the appropriate
   // vertex layouts, blend states, depth/stencil states, and shader modules.
   void CreateAllPipelines () {
      TextureFormat fmt = mGPU.SurfaceFormat;

      // --- 2D instanced line pipelines (Line2D, Bezier2D, DashLine2D) ---
      GPUVertexLayout[] line2DLayout = [InstanceLayout2DLine ()];
      Build (EPipeline.Line2D, "Line2D", line2DLayout, fmt, blend: true);
      Build (EPipeline.Bezier2D, "Bezier2D", line2DLayout, fmt, blend: true);
      Build (EPipeline.DashLine2D, "DashLine2D", line2DLayout, fmt, blend: true, textured: true);

      // --- 3D instanced line pipelines (Line3D, BlackLine, GlassLine) ---
      GPUVertexLayout[] line3DLayout = [InstanceLayout3DLine ()];
      Build (EPipeline.Line3D, "Line3D", line3DLayout, fmt, blend: true, depth: true);
      Build (EPipeline.BlackLine, "Line3D", line3DLayout, fmt, blend: true, depth: true);
      Build (EPipeline.GlassLine, "GlassLine", line3DLayout, fmt, blend: true, depth: true);

      // --- 2D instanced point pipeline ---
      GPUVertexLayout[] pt2DLayout = [InstanceLayout2DPoint ()];
      Build (EPipeline.Point2D, "Point2D", pt2DLayout, fmt, blend: true);

      // --- 3D instanced point pipeline ---
      GPUVertexLayout[] pt3DLayout = [InstanceLayout3DPoint ()];
      Build (EPipeline.Point3D, "Point3D", pt3DLayout, fmt, blend: true);

      // --- 2D direct triangle pipelines (Triangle2D, Quad2D) ---
      GPUVertexLayout[] flat2DLayout = [VertexLayout2D ()];
      Build (EPipeline.Triangle2D, "Flat2D", flat2DLayout, fmt);
      Build (EPipeline.Quad2D, "Flat2D", flat2DLayout, fmt);

      // --- 3D facet pipelines (Gourad, Phong, PhongPink, Pick, Glass, FlatFacet) ---
      GPUVertexLayout[] facet3DLayout = [VertexLayout3DFacet ()];
      Build (EPipeline.Gourad, "Gourad", facet3DLayout, fmt, depth: true);
      Build (EPipeline.Phong, "Phong", facet3DLayout, fmt, depth: true);
      Build (EPipeline.PhongPink, "PhongPink", facet3DLayout, fmt, depth: true);
      Build (EPipeline.Pick, "Pick", facet3DLayout, fmt, depth: true);
      Build (EPipeline.Glass, "Glass", facet3DLayout, fmt, depth: true);
      Build (EPipeline.FlatFacet, "FlatFacet", facet3DLayout, fmt, depth: true);

      // --- Instanced text pipelines ---
      GPUVertexLayout[] textPxLayout = [InstanceLayoutTextPx ()];
      Build (EPipeline.TextPx, "TextPx", textPxLayout, fmt, blend: true, textured: true);

      GPUVertexLayout[] text2DLayout = [InstanceLayoutText2D ()];
      Build (EPipeline.Text2D, "Text2D", text2DLayout, fmt, blend: true, textured: true);

      GPUVertexLayout[] text3DLayout = [InstanceLayoutText3D ()];
      Build (EPipeline.Text3D, "Text3D", text3DLayout, fmt, blend: true, depth: true, textured: true);

      // --- Stencil pipelines (TriFanStencil, TriFanCover) ---
      GPUStencilConfig stencilWrite = new () {
         FrontCompare = CompareFunction.Always,
         FrontPassOp = StencilOperation.Invert,
         FrontFailOp = StencilOperation.Keep,
         BackCompare = CompareFunction.Always,
         BackPassOp = StencilOperation.Invert,
         BackFailOp = StencilOperation.Keep,
         ReadMask = 0xFF,
         WriteMask = 0xFF
      };
      GPUStencilConfig stencilTest = new () {
         FrontCompare = CompareFunction.NotEqual,
         FrontPassOp = StencilOperation.Zero,
         FrontFailOp = StencilOperation.Keep,
         BackCompare = CompareFunction.NotEqual,
         BackPassOp = StencilOperation.Zero,
         BackFailOp = StencilOperation.Keep,
         ReadMask = 0xFF,
         WriteMask = 0xFF
      };
      BuildStencil (EPipeline.TriFanStencil, "Flat2D", flat2DLayout, fmt, stencilWrite,
         colorWrite: ColorWriteMask.None);
      BuildStencil (EPipeline.TriFanCover, "Flat2D", flat2DLayout, fmt, stencilTest,
         colorWrite: ColorWriteMask.All);
   }

   // Builds a standard pipeline with optional blend, depth, and texture bindings.
   void Build (EPipeline id, string shaderName, GPUVertexLayout[] layouts,
      TextureFormat fmt, bool blend = false, bool depth = false, bool textured = false) {

      GPUShaderModule shader = mShaders[shaderName];
      BindGroupLayout*[] bgl = textured
         ? [mTexturedLayout] : [mUniformLayout];
      mPipelines[id] = GPUPipeline.Create (
         mGPU, shader, "vs_main", shader, "fs_main",
         layouts, fmt,
         enableBlend: blend,
         enableDepth: depth,
         bindGroupLayouts: bgl,
         label: id.ToString ());
   }

   // Builds a stencil pipeline with custom stencil config and color write mask.
   void BuildStencil (EPipeline id, string shaderName, GPUVertexLayout[] layouts,
      TextureFormat fmt, GPUStencilConfig stencilCfg, ColorWriteMask colorWrite) {

      GPUShaderModule shader = mShaders[shaderName];
      BindGroupLayout*[] bgl = [mUniformLayout];
      mPipelines[id] = GPUPipeline.Create (
         mGPU, shader, "vs_main", shader, "fs_main",
         layouts, fmt,
         enableBlend: false,
         enableDepth: false,
         bindGroupLayouts: bgl,
         label: id.ToString (),
         stencil: stencilCfg,
         colorWriteMask: colorWrite);
   }

   // --- Vertex layout helpers ------------------------------------------------

   // 2D line instance: two vec2<f32> endpoints (p0, p1), step per instance
   static GPUVertexLayout InstanceLayout2DLine ()
      => new () {
         Stride = 16, StepMode = VertexStepMode.Instance,
         Attributes = [
            new () { Location = 0, Format = VertexFormat.Float32x2, Offset = 0 },
            new () { Location = 1, Format = VertexFormat.Float32x2, Offset = 8 }
         ]
      };

   // 3D line instance: two vec3<f32> endpoints (p0, p1), step per instance
   static GPUVertexLayout InstanceLayout3DLine ()
      => new () {
         Stride = 24, StepMode = VertexStepMode.Instance,
         Attributes = [
            new () { Location = 0, Format = VertexFormat.Float32x3, Offset = 0 },
            new () { Location = 1, Format = VertexFormat.Float32x3, Offset = 12 }
         ]
      };

   // 2D point instance: one vec2<f32> position, step per instance
   static GPUVertexLayout InstanceLayout2DPoint ()
      => new () {
         Stride = 8, StepMode = VertexStepMode.Instance,
         Attributes = [
            new () { Location = 0, Format = VertexFormat.Float32x2, Offset = 0 }
         ]
      };

   // 3D point instance: one vec3<f32> position, step per instance
   static GPUVertexLayout InstanceLayout3DPoint ()
      => new () {
         Stride = 12, StepMode = VertexStepMode.Instance,
         Attributes = [
            new () { Location = 0, Format = VertexFormat.Float32x3, Offset = 0 }
         ]
      };

   // 2D direct vertex: one vec2<f32> position, step per vertex
   static GPUVertexLayout VertexLayout2D ()
      => new () {
         Stride = 8, StepMode = VertexStepMode.Vertex,
         Attributes = [
            new () { Location = 0, Format = VertexFormat.Float32x2, Offset = 0 }
         ]
      };

   // 3D facet vertex: vec3<f32> position + vec3<f32> normal, step per vertex
   static GPUVertexLayout VertexLayout3DFacet ()
      => new () {
         Stride = 24, StepMode = VertexStepMode.Vertex,
         Attributes = [
            new () { Location = 0, Format = VertexFormat.Float32x3, Offset = 0 },
            new () { Location = 1, Format = VertexFormat.Float32x3, Offset = 12 }
         ]
      };

   // TextPx instance: vec4<i32> char_box + i32 tex_offset, step per instance
   // The Vec4S (4 shorts = 8 bytes) is expanded to vec4<i32> (16 bytes) on upload,
   // and the int tex_offset is 4 bytes → total stride 20 bytes.
   static GPUVertexLayout InstanceLayoutTextPx ()
      => new () {
         Stride = 20, StepMode = VertexStepMode.Instance,
         Attributes = [
            new () { Location = 0, Format = VertexFormat.Sint32x4, Offset = 0 },
            new () { Location = 1, Format = VertexFormat.Sint32, Offset = 16 }
         ]
      };

   // Text2D instance: vec2<f32> pos + vec4<i32> char_box + i32 tex_offset
   // 8 + 16 + 4 = 28 bytes, step per instance
   static GPUVertexLayout InstanceLayoutText2D ()
      => new () {
         Stride = 28, StepMode = VertexStepMode.Instance,
         Attributes = [
            new () { Location = 0, Format = VertexFormat.Float32x2, Offset = 0 },
            new () { Location = 1, Format = VertexFormat.Sint32x4, Offset = 8 },
            new () { Location = 2, Format = VertexFormat.Sint32, Offset = 24 }
         ]
      };

   // Text3D instance: vec3<f32> pos + vec4<i32> char_box + i32 tex_offset
   // 12 + 16 + 4 = 32 bytes, step per instance
   static GPUVertexLayout InstanceLayoutText3D ()
      => new () {
         Stride = 32, StepMode = VertexStepMode.Instance,
         Attributes = [
            new () { Location = 0, Format = VertexFormat.Float32x3, Offset = 0 },
            new () { Location = 1, Format = VertexFormat.Sint32x4, Offset = 12 },
            new () { Location = 2, Format = VertexFormat.Sint32, Offset = 28 }
         ]
      };

   // Private data -------------------------------------------------------------
   GPUDevice mGPU = null!;
   Dictionary<EPipeline, GPUPipeline> mPipelines = [];
   Dictionary<string, GPUShaderModule> mShaders = [];
   BindGroupLayout* mUniformLayout;
   BindGroupLayout* mTexturedLayout;
}
#endregion

File: Nori.Client/src/engine/pipeline-factory.ts
// pipeline-factory.ts -- Pre-compiles all WebGPU render pipelines at initialization
// Ports PipelineFactory.cs to the browser WebGPU API.

// Import shader sources as raw strings. These are loaded at build time
// or fetched at runtime. For now we embed them inline via a loader map.
import { shaderSources } from './shader-loader.js';

// ---------------------------------------------------------------------------
// EPipeline enum -- matches the C# EPipeline exactly
// ---------------------------------------------------------------------------
export enum Pipeline {
  Line2D = 0,
  Line3D = 1,
  Bezier2D = 2,
  DashLine2D = 3,
  Point2D = 4,
  Point3D = 5,
  Triangle2D = 6,
  Quad2D = 7,
  BlackLine = 8,
  GlassLine = 9,
  Gourad = 10,
  Phong = 11,
  PhongPink = 12,
  Pick = 13,
  Glass = 14,
  FlatFacet = 15,
  TextPx = 16,
  Text2D = 17,
  Text3D = 18,
  TriFanStencil = 19,
  TriFanCover = 20,
}

// ---------------------------------------------------------------------------
// Vertex layout helpers -- mirror the C# GPUVertexLayout factories
// ---------------------------------------------------------------------------

/** 2D line instance: two vec2<f32> endpoints (p0, p1), 16 bytes per instance */
function instanceLayout2DLine(): GPUVertexBufferLayout {
  return {
    arrayStride: 16,
    stepMode: 'instance',
    attributes: [
      { shaderLocation: 0, format: 'float32x2', offset: 0 },
      { shaderLocation: 1, format: 'float32x2', offset: 8 },
    ],
  };
}

/** 3D line instance: two vec3<f32> endpoints (p0, p1), 24 bytes per instance */
function instanceLayout3DLine(): GPUVertexBufferLayout {
  return {
    arrayStride: 24,
    stepMode: 'instance',
    attributes: [
      { shaderLocation: 0, format: 'float32x3', offset: 0 },
      { shaderLocation: 1, format: 'float32x3', offset: 12 },
    ],
  };
}

/** 2D point instance: one vec2<f32> position, 8 bytes per instance */
function instanceLayout2DPoint(): GPUVertexBufferLayout {
  return {
    arrayStride: 8,
    stepMode: 'instance',
    attributes: [
      { shaderLocation: 0, format: 'float32x2', offset: 0 },
    ],
  };
}

/** 3D point instance: one vec3<f32> position, 12 bytes per instance */
function instanceLayout3DPoint(): GPUVertexBufferLayout {
  return {
    arrayStride: 12,
    stepMode: 'instance',
    attributes: [
      { shaderLocation: 0, format: 'float32x3', offset: 0 },
    ],
  };
}

/** 2D direct vertex: one vec2<f32> position, 8 bytes per vertex */
function vertexLayout2D(): GPUVertexBufferLayout {
  return {
    arrayStride: 8,
    stepMode: 'vertex',
    attributes: [
      { shaderLocation: 0, format: 'float32x2', offset: 0 },
    ],
  };
}

/** 3D facet vertex: vec3<f32> position + vec3<f32> normal, 24 bytes per vertex */
function vertexLayout3DFacet(): GPUVertexBufferLayout {
  return {
    arrayStride: 24,
    stepMode: 'vertex',
    attributes: [
      { shaderLocation: 0, format: 'float32x3', offset: 0 },
      { shaderLocation: 1, format: 'float32x3', offset: 12 },
    ],
  };
}

/** TextPx instance: vec4<i32> char_box + i32 tex_offset, 20 bytes per instance */
function instanceLayoutTextPx(): GPUVertexBufferLayout {
  return {
    arrayStride: 20,
    stepMode: 'instance',
    attributes: [
      { shaderLocation: 0, format: 'sint32x4', offset: 0 },
      { shaderLocation: 1, format: 'sint32', offset: 16 },
    ],
  };
}

/** Text2D instance: vec2<f32> pos + vec4<i32> char_box + i32 tex_offset, 28 bytes */
function instanceLayoutText2D(): GPUVertexBufferLayout {
  return {
    arrayStride: 28,
    stepMode: 'instance',
    attributes: [
      { shaderLocation: 0, format: 'float32x2', offset: 0 },
      { shaderLocation: 1, format: 'sint32x4', offset: 8 },
      { shaderLocation: 2, format: 'sint32', offset: 24 },
    ],
  };
}

/** Text3D instance: vec3<f32> pos + vec4<i32> char_box + i32 tex_offset, 32 bytes */
function instanceLayoutText3D(): GPUVertexBufferLayout {
  return {
    arrayStride: 32,
    stepMode: 'instance',
    attributes: [
      { shaderLocation: 0, format: 'float32x3', offset: 0 },
      { shaderLocation: 1, format: 'sint32x4', offset: 12 },
      { shaderLocation: 2, format: 'sint32', offset: 28 },
    ],
  };
}

// ---------------------------------------------------------------------------
// Blend state used for anti-aliased primitives (lines, points, text)
// ---------------------------------------------------------------------------
const ALPHA_BLEND: GPUBlendState = {
  color: {
    srcFactor: 'src-alpha',
    dstFactor: 'one-minus-src-alpha',
    operation: 'add',
  },
  alpha: {
    srcFactor: 'one',
    dstFactor: 'one-minus-src-alpha',
    operation: 'add',
  },
};

// ---------------------------------------------------------------------------
// PipelineFactory
// ---------------------------------------------------------------------------
export class PipelineFactory {
  private pipelines: Map<Pipeline, GPURenderPipeline> = new Map();
  private shaderModules: Map<string, GPUShaderModule> = new Map();
  uniformLayout!: GPUBindGroupLayout;
  texturedLayout!: GPUBindGroupLayout;

  /** Initialize all pipelines for the given device and canvas format */
  async init(device: GPUDevice, format: GPUTextureFormat): Promise<void> {
    this.loadShaders(device);
    this.createBindGroupLayouts(device);
    this.createAllPipelines(device, format);
  }

  /** Retrieve a pre-compiled render pipeline by type */
  get(pipeline: Pipeline): GPURenderPipeline {
    const p = this.pipelines.get(pipeline);
    if (!p) throw new Error(`Pipeline ${Pipeline[pipeline]} not compiled`);
    return p;
  }

  /** Check if a pipeline exists */
  has(pipeline: Pipeline): boolean {
    return this.pipelines.has(pipeline);
  }

  // -------------------------------------------------------------------------
  // Internals
  // -------------------------------------------------------------------------

  private loadShaders(device: GPUDevice): void {
    const sources = shaderSources();
    for (const [name, code] of Object.entries(sources)) {
      this.shaderModules.set(name, device.createShaderModule({
        code,
        label: name,
      }));
    }
  }

  private createBindGroupLayouts(device: GPUDevice): void {
    // Uniform-only layout: binding 0 = uniform buffer (vertex+fragment)
    this.uniformLayout = device.createBindGroupLayout({
      label: 'uniform-layout',
      entries: [{
        binding: 0,
        visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
        buffer: { type: 'uniform' },
      }],
    });

    // Textured layout: binding 0 = uniform, binding 1 = texture, binding 2 = sampler
    this.texturedLayout = device.createBindGroupLayout({
      label: 'textured-layout',
      entries: [
        {
          binding: 0,
          visibility: GPUShaderStage.VERTEX | GPUShaderStage.FRAGMENT,
          buffer: { type: 'uniform' },
        },
        {
          binding: 1,
          visibility: GPUShaderStage.FRAGMENT,
          texture: { sampleType: 'float', viewDimension: '2d', multisampled: false },
        },
        {
          binding: 2,
          visibility: GPUShaderStage.FRAGMENT,
          sampler: { type: 'filtering' },
        },
      ],
    });
  }

  private createAllPipelines(device: GPUDevice, format: GPUTextureFormat): void {
    const line2DLayout = [instanceLayout2DLine()];
    this.build(device, format, Pipeline.Line2D, 'Line2D', line2DLayout, { blend: true });
    this.build(device, format, Pipeline.Bezier2D, 'Bezier2D', line2DLayout, { blend: true });
    this.build(device, format, Pipeline.DashLine2D, 'DashLine2D', line2DLayout, { blend: true, textured: true });

    const line3DLayout = [instanceLayout3DLine()];
    this.build(device, format, Pipeline.Line3D, 'Line3D', line3DLayout, { blend: true, depth: true });
    this.build(device, format, Pipeline.BlackLine, 'Line3D', line3DLayout, { blend: true, depth: true });
    this.build(device, format, Pipeline.GlassLine, 'GlassLine', line3DLayout, { blend: true, depth: true });

    const pt2DLayout = [instanceLayout2DPoint()];
    this.build(device, format, Pipeline.Point2D, 'Point2D', pt2DLayout, { blend: true });

    const pt3DLayout = [instanceLayout3DPoint()];
    this.build(device, format, Pipeline.Point3D, 'Point3D', pt3DLayout, { blend: true });

    const flat2DLayout = [vertexLayout2D()];
    this.build(device, format, Pipeline.Triangle2D, 'Flat2D', flat2DLayout, {});
    this.build(device, format, Pipeline.Quad2D, 'Flat2D', flat2DLayout, {});

    const facet3DLayout = [vertexLayout3DFacet()];
    this.build(device, format, Pipeline.Gourad, 'Gourad', facet3DLayout, { depth: true });
    this.build(device, format, Pipeline.Phong, 'Phong', facet3DLayout, { depth: true });
    this.build(device, format, Pipeline.PhongPink, 'PhongPink', facet3DLayout, { depth: true });
    this.build(device, format, Pipeline.Pick, 'Pick', facet3DLayout, { depth: true });
    this.build(device, format, Pipeline.Glass, 'Glass', facet3DLayout, { depth: true });
    this.build(device, format, Pipeline.FlatFacet, 'FlatFacet', facet3DLayout, { depth: true });

    const textPxLayout = [instanceLayoutTextPx()];
    this.build(device, format, Pipeline.TextPx, 'TextPx', textPxLayout, { blend: true, textured: true });

    const text2DLayout = [instanceLayoutText2D()];
    this.build(device, format, Pipeline.Text2D, 'Text2D', text2DLayout, { blend: true, textured: true });

    const text3DLayout = [instanceLayoutText3D()];
    this.build(device, format, Pipeline.Text3D, 'Text3D', text3DLayout, { blend: true, depth: true, textured: true });

    // Stencil pipelines for tri-fan fill
    this.buildStencil(device, format, Pipeline.TriFanStencil, 'Flat2D', flat2DLayout, {
      compare: 'always',
      passOp: 'invert',
      failOp: 'keep',
      colorWrite: 0x0,   // no color output
    });
    this.buildStencil(device, format, Pipeline.TriFanCover, 'Flat2D', flat2DLayout, {
      compare: 'not-equal',
      passOp: 'zero',
      failOp: 'keep',
      colorWrite: GPUColorWrite.ALL,
    });
  }

  private build(
    device: GPUDevice,
    format: GPUTextureFormat,
    id: Pipeline,
    shaderName: string,
    layouts: GPUVertexBufferLayout[],
    opts: { blend?: boolean; depth?: boolean; textured?: boolean },
  ): void {
    const shader = this.shaderModules.get(shaderName);
    if (!shader) {
      console.warn(`Shader "${shaderName}" not found, skipping pipeline ${Pipeline[id]}`);
      return;
    }
    const bindGroupLayout = opts.textured ? this.texturedLayout : this.uniformLayout;
    const pipelineLayout = device.createPipelineLayout({
      label: `${Pipeline[id]}-layout`,
      bindGroupLayouts: [bindGroupLayout],
    });

    const colorTarget: GPUColorTargetState = {
      format,
      ...(opts.blend ? { blend: ALPHA_BLEND } : {}),
      writeMask: GPUColorWrite.ALL,
    };

    // All pipelines must declare depth/stencil format to match the render pass.
    // 3D pipelines write depth; 2D pipelines pass through without depth testing.
    const depthStencil: GPUDepthStencilState = opts.depth
      ? {
        format: 'depth24plus-stencil8',
        depthWriteEnabled: true,
        depthCompare: 'less-equal',
      }
      : {
        format: 'depth24plus-stencil8',
        depthWriteEnabled: false,
        depthCompare: 'always',
      };

    const pipeline = device.createRenderPipeline({
      label: Pipeline[id],
      layout: pipelineLayout,
      vertex: {
        module: shader,
        entryPoint: 'vs_main',
        buffers: layouts,
      },
      fragment: {
        module: shader,
        entryPoint: 'fs_main',
        targets: [colorTarget],
      },
      primitive: {
        topology: 'triangle-list',
        cullMode: 'none',
      },
      depthStencil,
    });
    this.pipelines.set(id, pipeline);
  }

  private buildStencil(
    device: GPUDevice,
    format: GPUTextureFormat,
    id: Pipeline,
    shaderName: string,
    layouts: GPUVertexBufferLayout[],
    opts: {
      compare: GPUCompareFunction;
      passOp: GPUStencilOperation;
      failOp: GPUStencilOperation;
      colorWrite: GPUColorWriteFlags;
    },
  ): void {
    const shader = this.shaderModules.get(shaderName);
    if (!shader) return;
    const pipelineLayout = device.createPipelineLayout({
      label: `${Pipeline[id]}-layout`,
      bindGroupLayouts: [this.uniformLayout],
    });

    const stencilFace: GPUStencilFaceState = {
      compare: opts.compare,
      passOp: opts.passOp,
      failOp: opts.failOp,
      depthFailOp: 'keep',
    };

    const pipeline = device.createRenderPipeline({
      label: Pipeline[id],
      layout: pipelineLayout,
      vertex: {
        module: shader,
        entryPoint: 'vs_main',
        buffers: layouts,
      },
      fragment: {
        module: shader,
        entryPoint: 'fs_main',
        targets: [{
          format,
          writeMask: opts.colorWrite,
        }],
      },
      primitive: {
        topology: 'triangle-list',
        cullMode: 'none',
      },
      depthStencil: {
        format: 'depth24plus-stencil8',
        depthWriteEnabled: false,
        depthCompare: 'always',
        stencilFront: stencilFace,
        stencilBack: stencilFace,
        stencilReadMask: 0xFF,
        stencilWriteMask: 0xFF,
      },
    });
    this.pipelines.set(id, pipeline);
  }
}
